testing/test_tmpdir.py

```diff
@@ -445,3 +446,32 @@ def test(tmp_path):
     # running a second time and ensure we don't crash
     result = pytester.runpytest("--basetemp=tmp")
     assert result.ret == 0
+
+
+@pytest.mark.usefixtures("invalid_get_user")
+def test_tmpdir_factory_handles_invalid_dir_characters(
+    tmpdir_factory: TempdirFactory,
+) -> None:
+    assert "pytest-of-unknown" in str(tmpdir_factory.mktemp("test"))
+
+
+@pytest.mark.usefixtures("invalid_get_user")
+def test_tmp_path_factory_handles_invalid_dir_characters(
+    tmp_path_factory: TempPathFactory,
+) -> None:
+    assert "pytest-of-unknown" in str(tmp_path_factory.mktemp("test"))
+
+
+@pytest.mark.usefixtures("invalid_get_user")
+def test_tmpdir_handles_invalid_dir_characters(tmpdir) -> None:
+    assert "pytest-of-unknown" in str(tmpdir)
+
+
+@pytest.mark.usefixtures("invalid_get_user")
+def test_pytester_handles_invalid_dir_characters(pytester: Pytester) -> None:
+    assert "pytest-of-unknown" in str(pytester)
+
+
+@pytest.fixture
+def invalid_get_user(monkeypatch: MonkeyPatch) -> None:
+    monkeypatch.setattr("getpass.getuser", lambda: "os/agnostic")
```

Comment:
Cool, you might want to throw a few other characters there too:

```suggestion
    monkeypatch.setattr("getpass.getuser", lambda: "os/:?*agnostic")
```

Reply:
thanks, @nicoddemus - any idea why the monkeypatch here is applied when running a single test from pytest but not when running via `pytest testing/tmp_dir.py` or via `tox -e py38` etc?  Will update this after work later

Reply:
I think it might be because the base temp directory is cached? I suggest to manually reset the base temp directory inside the test, or better yet, just extract the functionality from `getbasetemp` into a function and unittest that.